# T4a Filtrado
**Autor/a:** Argail  
**Proyecto P1:** Análisis exploratorio del Titanic (numpy y pandas)

> Reglas: no renombres columnas, no uses bucles `for` sobre filas, comenta tu código.
> Revisa el contrato en el README del repositorio.

## 1. Carga de datos (todos)

In [1]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

## 2. Base común (solo P4, P5, P6 y P7)
> **P1, P2 y P3: borrad esta celda y la de código de abajo**, porque necesitáis los datos originales con sus nulos y duplicados.  
> **No la modifiquéis.** Aplica el tratamiento de nulos acordado en el contrato.

In [2]:
# --- BASE COMÚN (no modificar) ---
df["age"] = df["age"].fillna(df["age"].median())
df["embarked"] = df["embarked"].fillna(df["embarked"].mode()[0])
df["embark_town"] = df["embark_town"].fillna(df["embark_town"].mode()[0])
df = df.drop(columns="deck")

## 3. Mi tarea
> Voy a aplicar dos filtros con condiciones distintas y calcular para cada grupo el número de pasajeros y su porcentaje de supervivencia.

In [3]:
# Filtro 1: seleccionamos únicamente a las mujeres.
filtro_mujeres = df[df["sex"] == "female"]

# Filtro 2: seleccionamos únicamente a los pasajeros de primera clase.
filtro_primera_clase = df[df["pclass"] == 1]

# Estadísticas del filtro de mujeres.
num_mujeres = len(filtro_mujeres)
porcentaje_supervivencia_mujeres = filtro_mujeres["survived"].mean() * 100

# Estadísticas del filtro de primera clase.
num_primera_clase = len(filtro_primera_clase)
porcentaje_supervivencia_primera_clase = filtro_primera_clase["survived"].mean() * 100

# Mostramos los resultados de los dos filtros.
print("Mujeres:", num_mujeres)
print("Porcentaje de supervivencia de las mujeres:",
      round(porcentaje_supervivencia_mujeres, 2), "%")

print("Pasajeros de primera clase:", num_primera_clase)
print("Porcentaje de supervivencia de primera clase:",
      round(porcentaje_supervivencia_primera_clase, 2), "%")

Mujeres: 314
Porcentaje de supervivencia de las mujeres: 74.2 %
Pasajeros de primera clase: 216
Porcentaje de supervivencia de primera clase: 62.96 %


### Filtro combinando dos condiciones a la vez

Además de los dos filtros anteriores, combinamos dos condiciones en un mismo filtro con el operador `&` (y), para ver el efecto conjunto de sexo y clase.

In [4]:
# Filtro 3: mujeres Y de primera clase (las dos condiciones a la vez)
filtro_mujeres_primera = df[(df["sex"] == "female") & (df["pclass"] == 1)]

# Filtro 4: hombres adultos (18 años o más) de tercera clase
filtro_hombres_tercera = df[(df["sex"] == "male") & (df["pclass"] == 3) & (df["age"] >= 18)]

print("Mujeres de 1ª clase:", len(filtro_mujeres_primera),
      "- supervivencia:", round(filtro_mujeres_primera["survived"].mean() * 100, 2), "%")
print("Hombres adultos de 3ª clase:", len(filtro_hombres_tercera),
      "- supervivencia:", round(filtro_hombres_tercera["survived"].mean() * 100, 2), "%")

Mujeres de 1ª clase: 94 - supervivencia: 96.81 %
Hombres adultos de 3ª clase: 304 - supervivencia: 12.17 %


## 4. Hallazgo
En el dataset hay 314 mujeres y su porcentaje de supervivencia fue del 74,20%. En primera clase había 216 pasajeros y su porcentaje de supervivencia fue del 62,96%. Al combinar ambas condiciones, las 94 mujeres de 1ª clase sobrevivieron en un 96,81%, frente al 12,17% de los 304 hombres adultos de 3ª clase.